### PR TITLE
Add caching to unprivileged search results

### PR DESCRIPTION
**Pull request by mhilton**
Add a cache for search results for anonymous users.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1013/
Test PASSed.

File: internal/v4/api.go [hunk 1]
```diff
@@ -47,6 +48,7 @@ type Handler struct {
 	locator        *bakery.PublicKeyRing
 	identityClient *identity.Client
 	pool           *charmstore.Pool
+	searchCache    *cache.Cache
```
Comment (rogpeppe) [inline internal/v4/api.go:55]: Please add a comment saying what type the cache stores
and what it's keyed by.

File: internal/v4/common_test.go [hunk 2]
```diff
@@ -137,10 +137,11 @@ func (s *commonSuite) TearDownTest(c *gc.C) {
...
 func (s *commonSuite) startServer(c *gc.C) {
 	config := charmstore.ServerParams{
-		AuthUsername:     testUsername,
-		AuthPassword:     testPassword,
-		StatsCacheMaxAge: time.Nanosecond,
-		MaxMgoSessions:   s.maxMgoSessions,
+		AuthUsername:      testUsername,
+		AuthPassword:      testPassword,
+		StatsCacheMaxAge:  time.Nanosecond,
+		SearchCacheMaxAge: 0,
```
Comment (rogpeppe) [inline internal/v4/common_test.go]: Or just omit it?

File: internal/v4/search.go [hunk 3]
```diff
@@ -38,6 +38,20 @@ func (h *ReqHandler) serveSearch(_ http.Header, req *http.Request) (interface{},
...
 		sp.Groups = append(sp.Groups, groups...)
 	}
+	search := func() (interface{}, error) {
+		return h.doSearch(sp, req)
+	}
+	if sp.Admin || len(sp.Groups) > 0 {
+		// Don't use the cache if the request has elevated privileges.
+		return search()
+	}
+	return h.handler.searchCache.Get(req.URL.RawQuery, search)
```
Comment (rogpeppe) [inline internal/v4/search.go:48]: nice.

**Comment by rogpeppe** [comment]
LGTM assuming you've ensured that there's coverage of all the new code, thanks!

File: internal/v4/api.go (hunk 1 above)
Comment (frankban) [inline internal/v4/api.go:55]: +1

**Comment by frankban** [comment]
:+1:

File: internal/v4/api.go (hunk 1 above)
Comment (mhilton) [inline internal/v4/api.go:55]: done

File: internal/v4/common_test.go (hunk 2 above)
Comment (mhilton) [inline internal/v4/common_test.go]: done

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1014/
Test PASSed.

**Comment by mhilton** [comment]
:shipit: thanks for the reviews.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**